File: src/main/java/io/jenkins/plugins/gamekins/statistics/Statistics.java
package io.jenkins.plugins.gamekins.statistics;

import hudson.model.*;
import io.jenkins.plugins.gamekins.GameUserProperty;
import io.jenkins.plugins.gamekins.util.JacocoUtil;
import org.jenkinsci.plugins.workflow.job.WorkflowJob;
import org.jenkinsci.plugins.workflow.job.WorkflowRun;
import org.jenkinsci.plugins.workflow.multibranch.WorkflowMultiBranchProject;

import java.text.Collator;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Objects;
import java.util.Optional;

public class Statistics {

    private final int RUN_TOTAL_COUNT = 200;

    private final String projectName;
    private final ArrayList<RunEntry> runEntries;
    private boolean fullyInitialized;

    public Statistics(AbstractItem job) {
        this.projectName = job.getName();
        this.fullyInitialized = false;
        this.runEntries = generateRunEntries(job);
        this.fullyInitialized = true;
    }

    public boolean isNotFullyInitialized() {
        return !this.fullyInitialized || this.runEntries == null;
    }

    public String printToXML() {
        StringBuilder print = new StringBuilder();
        print.append("<Statistics project=\"").append(this.projectName).append("\">\n");
        ArrayList<User> users = new ArrayList<>(User.getAll());
        users.removeIf(user -> !user.getProperty(GameUserProperty.class).isParticipating(this.projectName));
        print.append("    <Users count=\"").append(users.size()).append("\">\n");
        for (User user : users) {
            GameUserProperty property = user.getProperty(GameUserProperty.class);
            if (property != null && property.isParticipating(this.projectName)) {
                print.append(property.printToXML(this.projectName, "        ")).append("\n");
            }
        }
        print.append("    </Users>\n");
        this.runEntries.removeIf(Objects::isNull);
        print.append("    <Runs count=\"").append(this.runEntries.size()).append("\">\n");
        for (RunEntry entry : this.runEntries) {
            print.append(entry.printToXML("        ")).append("\n");
        }
        print.append("    </Runs>\n");
        print.append("</Statistics>");
        return print.toString();
    }

    private ArrayList<RunEntry> generateRunEntries(AbstractItem job) {
        ArrayList<RunEntry> entries = new ArrayList<>();
        if (job instanceof WorkflowMultiBranchProject) {
            Optional<WorkflowJob> master = ((WorkflowMultiBranchProject) job).getItems().stream()
                    .filter(item -> item.getName().equals("master")).findFirst();
            if (master.isPresent()) {
                ArrayList<WorkflowRun> list = new ArrayList<>(master.get().getBuilds());
                Collections.reverse(list);
                for (WorkflowRun workflowRun : list) {
                    entries.add(new RunEntry(
                            workflowRun.getNumber(),
                            "master",
                            workflowRun.getResult(),
                            workflowRun.getStartTimeInMillis(),
                            0,
                            0,
                            JacocoUtil.getTestCount(null, workflowRun),
                            0.0));
                }
            } else {
                int count = 0;
                for (WorkflowJob workflowJob : ((WorkflowMultiBranchProject) job).getItems()) {
                    if (count >= RUN_TOTAL_COUNT) break;
                    ArrayList<WorkflowRun> list = new ArrayList<>(workflowJob.getBuilds());
                    Collections.reverse(list);
                    for (WorkflowRun workflowRun : list) {
                        if (count >= RUN_TOTAL_COUNT) break;
                        entries.add(new RunEntry(
                                workflowRun.getNumber(),
                                workflowJob.getName(),
                                workflowRun.getResult(),
                                workflowRun.getStartTimeInMillis(),
                                0,
                                0,
                                JacocoUtil.getTestCount(null, workflowRun),
                                0.0));
                        count++;
                    }
                }
            }
        } else if (job instanceof WorkflowJob) {
            ArrayList<WorkflowRun> list = new ArrayList<>(((WorkflowJob) job).getBuilds());
            Collections.reverse(list);
            for (WorkflowRun workflowRun : list) {
                entries.add(new RunEntry(
                        workflowRun.getNumber(),
                        "",
                        workflowRun.getResult(),
                        workflowRun.getStartTimeInMillis(),
                        0,
                        0,
                        JacocoUtil.getTestCount(null, workflowRun),
                        0.0));
            }
        } else if (job instanceof AbstractProject<?, ?>) {
            ArrayList<AbstractBuild<?, ?>> list = new ArrayList<>(((AbstractProject<?, ?>) job).getBuilds());
            Collections.reverse(list);
            for (AbstractBuild<?, ?> abstractBuild : list) {
                entries.add(new RunEntry(
                        abstractBuild.getNumber(),
                        "",
                        abstractBuild.getResult(),
                        abstractBuild.getStartTimeInMillis(),
                        0,
                        0,
                        JacocoUtil.getTestCount(null, abstractBuild),
                        0.0));
            }
        }
        entries.sort(RunEntry::compareTo);
        return entries;
    }

    public void addRunEntry(AbstractItem job, String branch, RunEntry entry, TaskListener listener) {
        addPreviousEntries(job, branch, entry.getRunNumber() - 1, listener);
        this.runEntries.add(entry);
        listener.getLogger().println(entry.printToXML(""));
        this.runEntries.sort(RunEntry::compareTo);
    }

    private void addPreviousEntries(AbstractItem job, String branch, int number, TaskListener listener) {
        if (number <= 0) return;
        for (RunEntry entry : this.runEntries) {
            if (entry.getBranch().equals(branch) && entry.getRunNumber() == number) return;
        }
        addPreviousEntries(job, branch, number - 1, listener);
        if (this.runEntries.size() > 0) {
            listener.getLogger().println(this.runEntries.get(this.runEntries.size() - 1).printToXML(""));
        }
        if (job instanceof WorkflowMultiBranchProject) {
            for (WorkflowJob workflowJob : ((WorkflowMultiBranchProject) job).getItems()) {
                if (workflowJob.getName().equals(branch)) {
                    ArrayList<WorkflowRun> list = new ArrayList<>(workflowJob.getBuilds());
                    for (WorkflowRun workflowRun : list) {
                        if (workflowRun.getNumber() == number) {
                            this.runEntries.add(new RunEntry(
                                    workflowRun.getNumber(),
                                    workflowJob.getName(),
                                    workflowRun.getResult(),
                                    workflowRun.getStartTimeInMillis(),
                                    0,
                                    0,
                                    JacocoUtil.getTestCount(null, workflowRun),
                                    0.0));
                            return;
                        }
                    }
                }
            }
        } else if (job instanceof WorkflowJob) {
            ArrayList<WorkflowRun> list = new ArrayList<>(((WorkflowJob) job).getBuilds());
            for (WorkflowRun workflowRun : list) {
                if (workflowRun.getNumber() == number) {
                    this.runEntries.add(new RunEntry(
                            workflowRun.getNumber(),
                            "",
                            workflowRun.getResult(),
                            workflowRun.getStartTimeInMillis(),
                            0,
                            0,
                            JacocoUtil.getTestCount(null, workflowRun),
                            0.0));
                    return;
                }
            }
        } else if (job instanceof AbstractProject<?, ?>) {
            ArrayList<AbstractBuild<?, ?>> list = new ArrayList<>(((AbstractProject<?, ?>) job).getBuilds());
            for (AbstractBuild<?, ?> abstractBuild : list) {
                if (abstractBuild.getNumber() == number) {
                    this.runEntries.add(new RunEntry(
                            abstractBuild.getNumber(),
                            "",
                            abstractBuild.getResult(),
                            abstractBuild.getStartTimeInMillis(),
                            0,
                            0,
                            JacocoUtil.getTestCount(null, abstractBuild),
                            0.0));
                    return;
                }
            }
        }
    }

    public static class RunEntry implements Comparable<RunEntry> {

        private final int runNumber;
        private final String branch;
        private final Result result;
        private final long startTime;
        private final int generatedChallenges;
        private final int solvedChallenges;
        private final int testCount;
        private final double coverage;

        public RunEntry(int runNumber, String branch, Result result, long startTime,
                        int generatedChallenges, int solvedChallenges, int testCount, double coverage) {
            this.runNumber = runNumber;
            this.branch = branch;
            this.result = result;
            this.startTime = startTime;
            this.generatedChallenges = generatedChallenges;
            this.solvedChallenges = solvedChallenges;
            this.testCount = testCount;
            this.coverage = coverage;
        }

        public int getRunNumber() {
            return runNumber;
        }

        public String getBranch() {
            return branch;
        }

        public Result getResult() {
            return result;
        }

        public long getStartTime() {
            return startTime;
        }

        public int getTestCount() {
            return testCount;
        }

        public double getCoverage() {
            return coverage;
        }

        public String printToXML(String indentation) {
            return indentation + "<Run number=\"" + this.runNumber + "\" branch=\"" + this.branch +
                    "\" result=\"" + (this.result == null ? "NULL" : this.result.toString()) + "\" startTime=\"" +
                    this.startTime + "\" generatedChallenges=\"" + this.generatedChallenges +
                    "\" solvedChallenges=\"" + this.solvedChallenges + "\" tests=\"" + this.testCount
                    + "\" coverage=\"" + this.coverage + "\"/>";
        }


        @Override
        public int compareTo(RunEntry o) {
            int result = Collator.getInstance().compare(this.getBranch(), o.getBranch());
            if (result == 0) return Integer.compare(this.getRunNumber(), o.getRunNumber());
            return result;
        }
    }
}

File: src/main/kotlin/io/jenkins/plugins/gamekins/statistics/Statistics.kt
package io.jenkins.plugins.gamekins.statistics

import hudson.model.*
import io.jenkins.plugins.gamekins.GameUserProperty
import io.jenkins.plugins.gamekins.util.JacocoUtil
import org.jenkinsci.plugins.workflow.job.WorkflowJob
import org.jenkinsci.plugins.workflow.multibranch.WorkflowMultiBranchProject
import java.text.Collator
import java.util.*

/**
 * Class for evaluation purposes. Displays the information about the users and the runs in an XML format.
 *
 * @author Philipp Straubinger
 * @since 1.0
 */
class Statistics(job: AbstractItem) {

    private var fullyInitialized: Boolean = false
    private val projectName: String = job.name
    private val runEntries: ArrayList<RunEntry>?

    init {
        runEntries = generateRunEntries(job)
        fullyInitialized = true
    }

    companion object {
        private const val RUN_TOTAL_COUNT = 200
    }

    /**
     * If Gamekins is not enabled when the [job] is created, this method adds the previous entries to the [Statistics]
     * according to the [branch] (only if [job] is of type [WorkflowMultiBranchProject]) recursively with the [number]
     * to the [runEntries]. It can happen that a job is aborted or fails and Gamekins is not executed. For this
     * problem, this method also adds old runs until the point where it last stopped. The [listener] reports the events
     * to the console output of Jenkins.
     */
    private fun addPreviousEntries(job: AbstractItem, branch: String, number: Int, listener: TaskListener) {
        if (number <= 0) return
        for (entry in runEntries!!) {
            if (entry.branch == branch && entry.runNumber == number) return
        }
        addPreviousEntries(job, branch, number - 1, listener)
        if (runEntries.size > 0) {
            listener.logger.println(runEntries[runEntries.size - 1].printToXML(""))
        }
        when (job) {
            is WorkflowMultiBranchProject -> {
                for (workflowJob in job.items) {
                    if (workflowJob.name == branch) {
                        for (workflowRun in workflowJob.builds) {
                            if (workflowRun.getNumber() == number) {
                                runEntries.add(RunEntry(
                                        workflowRun.getNumber(),
                                        workflowJob.name,
                                        workflowRun.result,
                                        workflowRun.startTimeInMillis,
                                        0,
                                        0,
                                        JacocoUtil.getTestCount(null, workflowRun),
                                        0.0))
                                return
                            }
                        }
                    }
                }
            }
            is WorkflowJob -> {
                for (workflowRun in job.builds) {
                    if (workflowRun.getNumber() == number) {
                        runEntries.add(RunEntry(
                                workflowRun.getNumber(),
                                "",
                                workflowRun.result,
                                workflowRun.startTimeInMillis,
                                0,
                                0,
                                JacocoUtil.getTestCount(null, workflowRun),
                                0.0))
                        return
                    }
                }
            }
            is AbstractProject<*, *> -> {
                for (abstractBuild in job.builds) {
                    if (abstractBuild.getNumber() == number) {
                        runEntries.add(RunEntry(
                                abstractBuild.getNumber(),
                                "",
                                abstractBuild.result,
                                abstractBuild.startTimeInMillis,
                                0,
                                0,
                                JacocoUtil.getTestCount(null, abstractBuild),
                                0.0))
                        return
                    }
                }
            }
        }
    }

    /**
     * Adds a new [entry] for a [job] and [branch] to the [Statistics]. Checks and optionally adds previous entries.
     * The [listener] reports the events to the console output of Jenkins.
     */
    fun addRunEntry(job: AbstractItem, branch: String, entry: RunEntry, listener: TaskListener) {
        addPreviousEntries(job, branch, entry.runNumber - 1, listener)
        runEntries!!.add(entry)
        listener.logger.println(entry.printToXML(""))
        runEntries.sortWith(Comparator { obj: RunEntry, o: RunEntry -> obj.compareTo(o) })
    }

    /**
     * Generates the [runEntries] during the creation of the the property of the [job]. Only adds the entries for the
     * branch master of a [WorkflowMultiBranchProject] to the [runEntries].
     */
    private fun generateRunEntries(job: AbstractItem): ArrayList<RunEntry> {
        val entries = ArrayList<RunEntry>()
        when (job) {
            is WorkflowMultiBranchProject -> {
                val master = job.items.stream().filter { item: WorkflowJob -> item.name == "master" }.findFirst()
                if (master.isPresent) {
                    val list = master.get().builds
                    list.reverse()
                    for (workflowRun in list) {
                        entries.add(RunEntry(
                                workflowRun!!.getNumber(),
                                "master",
                                workflowRun.result,
                                workflowRun.startTimeInMillis,
                                0,
                                0,
                                JacocoUtil.getTestCount(null, workflowRun),
                                0.0))
                    }
                } else {
                    var count = 0
                    for (workflowJob in job.items) {
                        if (count >= RUN_TOTAL_COUNT) break
                        val list = workflowJob.builds
                        list.reverse()
                        for (workflowRun in list) {
                            if (count >= RUN_TOTAL_COUNT) break
                            entries.add(RunEntry(
                                    workflowRun!!.getNumber(),
                                    workflowJob.name,
                                    workflowRun.result,
                                    workflowRun.startTimeInMillis,
                                    0,
                                    0,
                                    JacocoUtil.getTestCount(null, workflowRun),
                                    0.0))
                            count++
                        }
                    }
                }
            }
            is WorkflowJob -> {
                val list = job.builds
                list.reverse()
                for (workflowRun in list) {
                    entries.add(RunEntry(
                            workflowRun!!.getNumber(),
                            "",
                            workflowRun.result,
                            workflowRun.startTimeInMillis,
                            0,
                            0,
                            JacocoUtil.getTestCount(null, workflowRun),
                            0.0))
                }
            }
            is AbstractProject<*, *> -> {
                val list = job.builds
                list.reverse()
                for (abstractBuild in list) {
                    entries.add(RunEntry(
                            abstractBuild!!.getNumber(),
                            "",
                            abstractBuild.result,
                            abstractBuild.startTimeInMillis,
                            0,
                            0,
                            JacocoUtil.getTestCount(null, abstractBuild),
                            0.0))
                }
            }
        }
        entries.sortedWith(compareBy({it.branch}, {it.runNumber}))
        return entries
    }

    /**
     * If the [Statistics] is interrupted during initialization, it is triggered again.
     */
    fun isNotFullyInitialized(): Boolean {
        return !fullyInitialized || runEntries == null
    }

    /**
     * Returns an XML representation of the [Statistics].
     */
    fun printToXML(): String {
        val print = StringBuilder()
        print.append("<Statistics project=\"").append(projectName).append("\">\n")

        val users: ArrayList<User> = ArrayList(User.getAll())
        users.removeIf { user: User -> !user.getProperty(GameUserProperty::class.java).isParticipating(projectName) }
        print.append("    <Users count=\"").append(users.size).append("\">\n")
        for (user in users) {
            val property = user.getProperty(GameUserProperty::class.java)
            if (property != null && property.isParticipating(projectName)) {
                print.append(property.printToXML(projectName, "        ")).append("\n")
            }
        }
        print.append("    </Users>\n")

        runEntries!!.removeIf { obj: RunEntry? -> Objects.isNull(obj) }
        print.append("    <Runs count=\"").append(runEntries.size).append("\">\n")
        for (entry in runEntries) {
            print.append(entry.printToXML("        ")).append("\n")
        }
        print.append("    </Runs>\n")

        print.append("</Statistics>")
        return print.toString()
    }

    /**
     * Represents a run of Jenkins job.
     *
     * @author Philipp Straubinger
     * @since 1.0
     */
    class RunEntry(val runNumber: Int, val branch: String, val result: Result?, val startTime: Long,
                   private val generatedChallenges: Int, private val solvedChallenges: Int, val testCount: Int,
                   val coverage: Double)
        : Comparable<RunEntry> {

        /**
         * Returns an XML representation of the [RunEntry].
         */
        fun printToXML(indentation: String): String {
            return (indentation + "<Run number=\"" + runNumber + "\" branch=\"" + branch +
                    "\" result=\"" + (result?.toString() ?: "NULL") + "\" startTime=\"" +
                    startTime + "\" generatedChallenges=\"" + generatedChallenges +
                    "\" solvedChallenges=\"" + solvedChallenges + "\" tests=\"" + testCount
                    + "\" coverage=\"" + coverage + "\"/>")
        }

        /**
         * Compares two [RunEntry] with first [branch] and second [runNumber].
         *
         * @see [Comparable.compareTo]
         */
        override fun compareTo(other: RunEntry): Int {
            val result = Collator.getInstance().compare(branch, other.branch)
            return if (result == 0) runNumber.compareTo(other.runNumber) else result
        }
    }
}
